README
Pkgtxt2db is a utility for Slackware and Slackware based operating system only.

It can be use to convert PACKAGES.TXT to various database formats:
Are supported:  - CSV
                - JSON
                - XML


Download:
---------
Sources tarballs and Slackware packages can be found there:
http://people.salixos.org/fredg/packages/Pkgtxt2db/

GIT:
----
git clone ssh://git@github.com/fredg/pkgtxt2db.git


------------------------------------------------------------------
GENERAL OPTIONS
       -h, --help
              Simple help

       --version
              Print pkgtxt2db version

       -u, --update
              It will fetch and download a new PACKAGES.TXT from your choosen repository.

       -t, --target
              Choose between slackware or salix repositories.

       --repo Choose the arch repository, it can be i486 or x86_64.

       -r, --release
              Choose the slackware or salix release.  It can be: current (if exist), 14.0, 13.37, ..., 13.0.

       -e, --expa
              Choose between the patches or extra Slackware repository.

       -c, --convert
              Choose your database format, it can be:

              ·  csv : Convert PACKAGES.TXT to a CSV database.

              ·  json: Convert PACKAGES.TXT to a JSON database.

              ·  xml : Convert PACKAGES.TXT to a XML database.

       -o, --output
              Choose the name of the output file.  If none is choosen, packages will be used.

COMMAND LINE USAGE
       Here are some usages examples.

       pkgtxt2db -u -t slackware --repo x86_64 -r current -c json -o slack64.json
              It will fetch and convert PACKAGES.TXT from the Slackware current x86_64 repository to a JSON database: slack64.json.

       pkgtxt2db -c csv
              It will convert the PACKAGES.TXT found in your directory to a CSV database (; separated): packages.csv.  You can open it with libreoffice ;)


ChangeLog
* version 0.2.2
	- CSV bugs fixes
	- Add missing values (suggested deps and conflicted pkgs)
	- Rewritten to use the ParsePkgtxt Class

* version 0.2.1
	- some code cleanup
	- add SQLite support

* version 0.2.0
	- code converted from Perl to Python
	- add support for the Slackware extra and patches directories
	- you can choose your output filename

* version 0.1.6
	- add XML export
	- clean pkgdesc var with parasite characters < > " &

* version 0.1.4
	- added pkgdesc to CSV DB
	- added "" delimiters to CSV, safer for mysql import
	- remove garbages from the Slackware DB output
	- fix pkgdesc to be one-line var
	- added output examples in /examples

* version 0.1.2
	- code cleanup
	- options are now human readable
	- man page added

* version 0.1.1
	- add pkgdesc var (package description) to JSON DB

* version 0.1
	- first try
	- Convert Salix and Slackware PACKAGES.TXT files to CSV and JSON DB.


Pkgtxt2db.pm
package Pkgtxt2db;

#
# Pkgtxt2db.pm
#
# This file is part of pkgtxt2db
# copyright 2011 Frédéric Galusik <fredg~at~salixos~dot~org>
#
# License: BSD Revised
#
use 5.010;
use strict;
use warnings;

use LWP::Simple;

our $pkgtxt = "PACKAGES.TXT";
our @releases = ("13.37","current");
our $release;
our $url;
our $target;
our @targets = qw/salix32 salix64 slack32 slack64/;
our %pkgdb;
#
# Salix vars
#
our $salix32 = "salix32";
our $salix64 = "salix64";
our $salix_mirror = "http://salix.enialis.net";
#
# Slackware vars
#
our $slack32 = "slack32";
our $slack64 = "slack64";
our $slack_mirror="http://slackware.org.uk/slackware";
#

sub new {
    my $self = {};
    bless ($self);
    return $self;
}

#
# choose the release
#
sub checkrelease {
    if ($release ~~ @releases) {
        print "Release: $release\n";
    } else {
        die "Not a valid release, see --help\n";
        }
    }

#
# choose the target between salix or slackware, i486 or x86_64
#
sub checktarget {
    if ($target ~~ @targets) {
        print "Target: $target\n";
    } else {
        die "You must choose a valid target, see --help\n";
    }
}

#
# build an array with all file lines
#
# Get the PACKAGES.TXT file
# Pkgtxt2db->getpkgtxt()
#
sub getpkgtxt {
    if ($target eq $salix32) {
        $url = "$salix_mirror/i486/$release/$pkgtxt";
    } elsif ($target eq $salix64) {
        $url = "$salix_mirror/x86_64/$release/$pkgtxt";
    } elsif ($target eq $slack32) {
        $url = "$slack_mirror/slackware-${release}/$pkgtxt";
    } elsif ($target eq $slack64) {
        $url = "$slack_mirror/slackware64-${release}/$pkgtxt";
    }
    print "URL: $url\n";
    my $status = getstore($url,$pkgtxt);
    if ( is_success($status) ) {
        print "PACKAGES.TXT downloaded correctly.\n";
    } else {
        print "PACKAGES.TXT not downloaded: $status\n";
    }
}


#
# turn PACKAGES.TXT to a array
#
sub getdata {
    open F, "<$pkgtxt" or die "No PACKAGES.TXT file, aborting.\n";
    our @data = <F>;
    close (F);
    return @data;
}

#
# Make the hash of array database
#
# build the DB
# Pkgtxt2db->mkdadb()
#
sub mkdadb {
    our $self = shift;

    our @d = getdata();
    our $pname;
    our $pkgname;

    foreach (@d) {
        chomp $_;
        if ($_ =~ /^$/){
            next;
        }
        if ($_ =~ /(^PACKAGE NAME:\s\s)(.*)/) {
            $pname = "$2";
            $pname =~ /^(.*)-([^-]*)-([^-]*)-([^-]*).t[glx]z$/;
            $pkgname = "$1";
            $pkgdb{$pkgname}[0] = "$pname";
            $pkgdb{$pkgname}[1] = "$2";
            $pkgdb{$pkgname}[2] = "$3";
            $pkgdb{$pkgname}[3] = "$4";
            next;
        }
        if ($_ =~ /(^PACKAGE LOCATION:\s\s\.)(.*)/) {
            $pkgdb{$pkgname}[4] = "$2";
            next;
        }
        if ($_ =~ /(^PACKAGE REQUIRED:\s\s)(.*)/) {
            $pkgdb{$pkgname}[5] = "$2";
            next;
        }
        if ($_ =~ /(^PACKAGE\sSIZE\s\(compressed\):\s\s)(.*)/) {
            $pkgdb{$pkgname}[6] = "$2";
            next;
        }
        if ($_ =~ /(^PACKAGE\sSIZE\s\(uncompressed\):\s\s)(.*)/) {
            $pkgdb{$pkgname}[7] = "$2";
            next;
        }
        if (($_ =~ /\Q${pkgname}\E:(.*)/) && (defined($pkgname))) {
            $pkgdb{$pkgname}[8] = $pkgdb{$pkgname}[8] . "$1";
            # clean space
            $pkgdb{$pkgname}[8] =~ s/ +/ /g;
            # clean xml output: & < > cannot be in a varchar
            $pkgdb{$pkgname}[8] =~ s/&/and/g;
            $pkgdb{$pkgname}[8] =~ s/(<)|(>)//g;
            # clean csv output: " in pkgdesc is shit
            $pkgdb{$pkgname}[8] =~ s/"/'/g;
            next;
        }
        else {
            next;
        }
            return %pkgdb;
    }
}

#
# CSV
#
sub tocsv {
    open(C, ">$target-$release.csv") or die "Unable to open $target-$release.csv for writing, aborting.";
    # choose the CSV separator, \t = tab \@ = @ ....
    # avoid the comma (,) as it is the separator for dependancies
    my $c = "\"\;\"";
    print C "\"pkgname${c}pkgver${c}arch${c}pkgrel${c}location${c}dep${c}sizeC${c}sizeU${c}Desc\"\n";
    for my $p ( sort keys %pkgdb ) {
        printf C "\"%s$c%s$c%s$c%s$c%s$c%s$c%s$c%s$c%s\"\n",
        $p, $pkgdb{$p}[1], $pkgdb{$p}[2], $pkgdb{$p}[3], $pkgdb{$p}[4], $pkgdb{$p}[5], $pkgdb{$p}[6], $pkgdb{$p}[7], $pkgdb{$p}[8];
    }
    close (C);
    print "$target-$release.csv has been built.\n";
}

#
# JSON
#
sub tojson {
    my $out = "$target-$release.json";
    my $n = keys %pkgdb;
    open(J, ">$out") or die "Unable to open $out for writing, aborting.";
    print J "\[\n";
    for my $p ( (keys %pkgdb)[0..($n-2)] ) {
        print J "  \{\n";
        print J "    \"pkgname\": \"$p\",\n";
        print J "    \"pkgver\": \"$pkgdb{$p}[1]\",\n";
        print J "    \"arch\": \"$pkgdb{$p}[2]\",\n";
        print J "    \"pkgver\": \"$pkgdb{$p}[3]\",\n";
        print J "    \"location\": \"$pkgdb{$p}[4]\",\n";
        print J "    \"dep\": \"$pkgdb{$p}[5]\",\n";
        print J "    \"sizeC\": \"$pkgdb{$p}[6]\",\n";
        print J "    \"sizeU\": \"$pkgdb{$p}[7]\"\n";
        print J "    \"pkgdesc\: \"$pkgdb{$p}[8]\"\n";
        print J "  \},\n";
    }
    for my $p ( (keys %pkgdb)[($n-1)] ) {
        print J "  \{\n";
        print J "    \"pkgname\": \"$p\",\n";
        print J "    \"pkgver\": \"$pkgdb{$p}[1]\",\n";
        print J "    \"arch\": \"$pkgdb{$p}[2]\",\n";
        print J "    \"pkgver\": \"$pkgdb{$p}[3]\",\n";
        print J "    \"location\": \"$pkgdb{$p}[4]\",\n";
        print J "    \"dep\": \"$pkgdb{$p}[5]\",\n";
        print J "    \"sizeC\": \"$pkgdb{$p}[6]\",\n";
        print J "    \"sizeU\": \"$pkgdb{$p}[7]\"\n";
        print J "    \"pkgdesc\: \"$pkgdb{$p}[8]\"\n";
        print J "  \}\n";
    }
    print J "\]\n";
    close (J);
    print "$target-$release.json has been built.\n";
}

sub toxml {
    my $outx = "$target-$release.xml";
    open(X, ">$outx") or die "Unable to open $outx for writing, aborting.";
    print X "<?xml version=\"1.0\" encoding=\"UTF-8\" ?>\n";
    print X "<packages>\n";
    for my $p (keys %pkgdb) {
        print X "\t<package>\n";
        print X "\t\t<pkgname>$p</pkgname>\n";
        print X "\t\t<pkgver>$pkgdb{$p}[1]</pkgver>\n";
        print X "\t\t<arch>$pkgdb{$p}[2]</arch>\n";
        print X "\t\t<pkgver>$pkgdb{$p}[3]</pkgver>\n";
        print X "\t\t<location>$pkgdb{$p}[4]</location>\n";
        print X "\t\t<dep>$pkgdb{$p}[5]</dep>\n";
        print X "\t\t<sizeC>$pkgdb{$p}[6]</sizeC>\n";
        print X "\t\t<sizeU>$pkgdb{$p}[7]</sizeU>\n";
        print X "\t\t<pkgdesc>$pkgdb{$p}[8]</pkgdesc>\n";
        print X "\t</package>\n";
    }
    print X "</packages>\n";
    close (X);
    print "$target-$release.xml has been built.\n";
}

1;


pkgtxt2db.pl
#!/usr/bin/env perl
#
# spkg2db.pl
#
# copyright 2011 Frédéric Galusik <fredg~at~salixos~dot~org>
#
# License: BSD Revised
#
# This perl module will convert the Slackware/Salix PACKAGES.TXT file
# to some various database format : CSV,
#

use strict;
use warnings;

use LWP::Simple;

#
# Get the PACKAGES.TXT file
# 
#
# Salix
our $salix_mirror = "http://salix.enialis.net";
our $arch32 = "i486";
our $arch64 = "x86_64";
our $release = "13.37";
our $pkgtxt = "/tmp/salixPACKAGES.TXT";
our $url32 = "$salix_mirror/$arch32/$release/PACKAGES.TXT";
our $url64 = "$salix_mirror/$arch64/$release/PACKAGES.TXT";
# Slackware
our $slack_mirror="ftp://ftp.osuosl.org/pub/slackware";
our $a32 = "slackware-";
our $a64 = "slackware64-";
our $slack32url = "$slack_mirror/$a32$release/PACKAGES.TXT";
our $slack64url = "$slack_mirror/$a64$release/PACKAGES.TXT";
our $slackpkgtxt = "/tmp/slackwarePACKAGES.TXT";

# Pkgtxt2db->get_file($url,$pkgtxt)
sub get_file {
    our ($a,$b) = @_;
    mirror($a, $b) or die "Error: could not fetch $a\n";
}

# 
# make the file an array
#
# Salix
# Pkgtxt2db->salix_data()
sub salix_data {
    open F, "<$pkgtxt" or die $!;
    our @salix_d = <F>;
    close (F);
    return @salix_d;
}
# Slackware
# Pkgtxt2db->slack_data()
sub slack_data {
    open G, "<$slackpkgtxt" or die $!;
    our @slack_d = <G>;
    close (G);
    return @slack_d;
}


#
# Make the hash of array database
our %pkgdb;

sub mkdadb {
    our $self = shift;

    our @d = salix_data() unless defined(@d);
    our $pname;
    our $pkgname;
   
    foreach (@d) {
        chomp $_;
        if ($_ =~ /^$/){
            next;
        }
        if ($_ =~ /(^PACKAGE NAME:\s\s)(.*)/) {
            $pname = "$2";
            $pname =~ /^(.*)-([^-]*)-([^-]*)-([^-]*).t[glx]z$/;
            $pkgname = "$1";
            $pkgdb{$pkgname}[0] = "$pname";
            $pkgdb{$pkgname}[1] = "$2";
            $pkgdb{$pkgname}[2] = "$3";
            $pkgdb{$pkgname}[3] = "$4";
            next;
        }
        if ($_ =~ /(^PACKAGE LOCATION:\s\s\.)(.*)/) {
            $pkgdb{$pkgname}[4] = "$2";
            next;
        }
        if ($_ =~ /(^PACKAGE REQUIRED:\s\s)(.*)/) {
            $pkgdb{$pkgname}[5] = "$2";
            next;
        }
        if ($_ =~ /(^PACKAGE\sSIZE\s\(compressed\):\s\s)(.*)/) {
            $pkgdb{$pkgname}[6] = "$2";
            next;
        }
        if ($_ =~ /(^PACKAGE\sSIZE\s\(uncompressed\):\s\s)(.*)/) {
            $pkgdb{$pkgname}[7] = "$2";
            next;
        }
        else {
            next;
        }
        if (@d = slack_data()) {
            our %slackdb = %pkgdb;
            return \%slackdb;
        } else {
            return %pkgdb;
        }
    }
}

sub test {
    get_file($url32,$pkgtxt);
    salix_data();
    mkdadb();
}
test();

# 
# CSV
# pkgname version arch release location dep sizeC sizeU
sub salix2csv {
    for our $p ( keys %pkgdb ) {
        print "$p\@$pkgdb{$p}[1]\@$pkgdb{$p}[2]\@$pkgdb{$p}[3]\@$pkgdb{$p}[4]\@$pkgdb{$p}[5]\@$pkgdb{$p}[6]\@$pkgdb{$p}[7]\n";
    }
}
salix2csv();



pkgtxt2db2.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
#
#       pkgtxt2db.py
#
#       Copyright 2012 Frédéric Galusik <fredg~at~salixos~dot~org>
#
#
# License: BSD Revised
#
# Convert the Slackware/Salix PACKAGES.TXT file to some various database
# formats : CSV, json, xml
#


import os
import sys
import urllib2
import gzip
import re
import argparse


# Program information
my_url = 'http://www.salixos.org/wiki/index.php/Pkgtxt2db'
my_name = 'pkgtxt2db'
my_version = '0.0'


def main():
# Parse the CLI options
    parser = argparse.ArgumentParser(
            prog='pkgtxt2db',
            description='Convert PACKAGES.TXT to DB',
            epilog=
            "i.e. pkgtxt2db -u -t salix -a x86_64 -r 14.0 -c json -o salix64.json")

    parser.add_argument('-u', '--update', action="store_true",
            default=False,
            help='Download/update the PACKAGES.TXT file')

    parser.add_argument('-t', '--target', action="store",
            dest='target', default='salix',
            help='Choose the O.S.: slackware or salix (default) ')

    parser.add_argument('--repo', action="store",
            dest='repo', default='i486',
            help='Choose the arch repo: x86_64 or i486 (default)')

    parser.add_argument('-e', '--expa', action="store",
            dest='expa', default='/',
            help='Choose the slackware extra/patches')

    parser.add_argument('-r', '--release', action="store",
            dest='release', default='14.0',
            help='Choose the release: 13.0 to 14.0 (default)')

    parser.add_argument('-c', '--convert', action="store",
            dest='convert',
            help='Choose the DB format: xml, json, csv')

    parser.add_argument('-o', '--output', action="store",
            dest='output', default='packages',
            help='Choose the name of your DB file')

    parser.add_argument('--version', action='version',
            version='%(prog)s ' + my_version)

    if len(sys.argv) == 1:
        sys.exit('Wrong usage, see pkgtxt2db --help')

    args = parser.parse_args()

    if not update and not os.path.isfile(pkgtxt):
            sys.exit('No PACKAGES.TXT found, you should fetch one, aborting.')
    else:
        pkgtxturl(repo, target, release, expa)

    if convert == 'csv':
        mkdadb(tocsv)
    elif convert == 'json':
        mkdadb(tojson)
    elif convert == 'xml':
        mkdadb(toxml)
    else:
        sys.exit('You have to choose a valid database format, aborting.')



    # vars
    mirror = 'http://salix.enialis.net/'
    pkgtxtz = 'PACKAGES.TXT.gz'
    pkgtxt = 'PACKAGES.TXT'
    update = args.update
    target = args.target
    repo = args.repo
    release = args.release
    expa = args.expa
    convert = args.convert
    output = args.output
    outputfile = '.'.join([output, convert])


# initialise the pkg dictionnary with empty values
fields = ['name',
             'version',
             'arch',
             'release'
             'location',
             'deps',
             'sizec',
             'sizeu',
             'slackdesc']


def new_pkgdct():
    return dict(zip(fields, ['', ] * len(fields)))


# Fetch PACKAGES.TXT
def pkgtxturl(repo='i486', target='salix', release='14.0', expa='/'):
    """
    Download the slackware/salix PACKAGES.TXT.gz from a built URL and unzip it.
    pkgtxturl(repo, target, release, |extra|patches)
    """

    # Check the repo
    Rep = {'i486', 'x86_64'}
    if repo not in Rep:
        sys.exit('Choose a valid arch please, aborting.')

    # Check if the choosen release is a valid one.
    R = {'13.0', '13.1', '13.37', '14.0', 'current'}
    if release not in R:
        sys.exit('Choose a valid release please, aborting.')

    # Check the target
    if target == 'slackware':
        target = 'slackware-'
    elif target == 'salix':
        target = ''
    else:
        sys.exit('Choose a valid target, aborting.')

    #
    if expa == 'extra':
        expa = '/extra/'
    elif expa == 'patches':
        expa = '/patches/'

    slash = '/'

    # Build the URL to fetch PACKAGES.TXT
    url = mirror + repo + slash + target + release + expa + pkgtxtz

    # remove old files
    if os.path.isfile(pkgtxtz):
        os.remove(pkgtxtz)
        print "Remove old ", pkgtxtz
    if os.path.isfile(pkgtxt):
        os.remove(pkgtxt)
        print "Remove old ", pkgtxt
    try:
        f = urllib2.urlopen(url)
        print "Fetching ", url
        print ""
        # Open local_file for writing
        with open(os.path.basename(url), "wb") as local_file:
            local_file.write(f.read())
    except urllib2.HTTPError, e:
        print "HTTP Error:", e.code, url
        return False
    except urllib2.URLError, e:
        print "URL Error:", e.reason, url
        return False

    # unzip it
    fout = open(pkgtxt, 'w')
    with gzip.open(pkgtxtz, 'rb') as f:
        for line in f:
            fout.write(line)
    fout.close()


# to CSV DB
def tocsv(pkgDct, sep=";"):
    """
    Export PACKAGES.TXT to a CSV database format.
    The separated string can be choosen with the sep var, default is ;
    """
    with open(outputfile, 'a') as csvf:
        csvf.write(
            sep.join(map(lambda field: pkgDct.get(field, ''), fields)) + '\n')


# to JSON DB
def tojson(pkgDct):
    """
    Export PACKAGES.TXT to a JSON database format
    """
    with open("pre.json", 'a') as j:
        j.write('  {\n')
        j.write('    \"name\": \"' + pkgDct.get("name") + '\",\n')
        j.write('    \"version\": \"' + pkgDct.get("version") + '\",\n')
        j.write('    \"arch\": \"' + pkgDct.get("arch") + '\",\n')
        j.write('    \"release\": \"' + pkgDct.get("release") + '\",\n')
        j.write('    \"location\": \"' + pkgDct.get("location") + '\",\n')
        j.write('    \"deps\": \"' + pkgDct.get("deps") + '\",\n')
        j.write('    \"sizec\": \"' + pkgDct.get("sizec") + '\",\n')
        j.write('    \"sizeu\": \"' + pkgDct.get("sizeu") + '\",\n')
        j.write('    \"slackdesc\": \"' + pkgDct.get("slackdesc") + '\"\n')
        j.write('  },\n')


# to XML DB
def toxml(pkgDct):
    """
    Export PACKAGES.TXT to a XML database format.
    """
    with open(outputfile, 'a') as xmlf:
        xmlf.write('\t<package>\n')
        xmlf.write('\t\t<name>' + pkgDct.get("name") + '</name>\n')
        xmlf.write('\t\t<version>' + pkgDct.get("version") + '</version>\n')
        xmlf.write('\t\t<arch>' + pkgDct.get("arch") + '</arch>\n')
        xmlf.write('\t\t<release>' + pkgDct.get("release") + '</release>\n')
        xmlf.write('\t\t<location>' + pkgDct.get("location") + '</location>\n')
        xmlf.write('\t\t<deps>' + pkgDct.get("deps") + '</deps>\n')
        xmlf.write('\t\t<sizec>' + pkgDct.get("sizec") + '</sizec>\n')
        xmlf.write('\t\t<sizeu>' + pkgDct.get("sizeu") + '</sizeu>\n')
        xmlf.write('\t\t<slackdesc>' + pkgDct.get("slackdesc") + '</slackdesc>\n')
        xmlf.write('\t</package>\n')


# parser
def mkdadb(towhat):
    """
    Parse PACKAGES.TXT to get the values we need.
    Choose the export format:
        - CSV  : tocsv
        - JSON : tojson
        - XML  : toxml
    """
    if towhat == tocsv:
        if os.path.isfile(outputfile):
            os.remove(outputfile)
            print outputfile, 'has been updated.'
        else:
            print outputfile, 'has been built.'
    if towhat == tojson:
        if os.path.isfile(outputfile):
            os.remove(outputfile)
            print outputfile, 'has been updated.'
        else:
            print outputfile, 'has been built.'
        with open("pre.json", 'w') as j:
                j.write('{\n')
                j.write('"packages": [\n')
    if towhat == toxml:
        if os.path.isfile(outputfile):
            os.remove(outputfile)
            print outputfile, 'has been updated.'
        else:
            print outputfile, 'has been built.'
        with open(outputfile, 'w') as xmlf:
                xmlf.write('<?xml version=\"1.0\" encoding=\"UTF-8\" ?>\n')
                xmlf.write('<packages>\n')
    pkg = new_pkgdct()
    with open('PACKAGES.TXT') as f:
        for line in f:
            pkgline = re.match(
                r'(PACKAGE NAME:\s\s)(.*)', line)
            locationline = re.match(
                r'(PACKAGE LOCATION:\s\s\.)(.*)', line)
            depline = re.match(
                r'(PACKAGE REQUIRED:\s\s)(.*)', line)
            sizecline = re.match(
                r'(PACKAGE\sSIZE\s\(compressed\):\s\s)(.*)', line)
            sizeuline = re.match(
                r'(PACKAGE\sSIZE\s\(uncompressed\):\s\s)(.*)', line)
            slackdescline = re.match(
                r'(%s:\s)(.*)' % pkg["name"].replace('+', '\+'), line)
            emptyline = re.match(
                r'^$', line)
            if pkgline:
                pname = pkgline.group(2)
                pname = re.match(
                    r'(.*)-([^-]*)-([^-]*)-([^-]*).t[glx]z$', pname)
                pkg["name"] = pname.group(1)
                pkg["version"] = pname.group(2)
                pkg["arch"] = pname.group(3)
                pkg["release"] = pname.group(4)
            if depline:
                pkg["deps"] = depline.group(2)
            if locationline:
                pkg["location"] = locationline.group(2)
            if sizecline:
                pkg["sizec"] = sizecline.group(2)
            if sizeuline:
                pkg["sizeu"] = sizeuline.group(2)
            if slackdescline:
                pkg["slackdesc"] += " " + slackdescline.group(2).\
                    replace('"', '\'').\
                    replace('&', 'and').\
                    replace('>', '').\
                    replace('<', '')
            if emptyline and pkg.get("name"):
                pkg["slackdesc"] = pkg["slackdesc"].strip()
                towhat(pkg)
                pkg = new_pkgdct()
    if towhat == tojson:
        with open("pre.json", 'r') as j, open(outputfile, "w") as jsonf:
            alllines = j.readlines()
            alllines[-1] = alllines[-1].replace('},', '}')
            jsonf.writelines(alllines)
            jsonf.write(']\n')
            jsonf.write('}\n')
            os.remove("pre.json")
    if towhat == toxml:
        with open(outputfile, 'a') as xmlf:
            xmlf.write('</packages>\n')


if __name__ == '__main__':
    main()
